### DISCUSSION TITLE
Keep rust code unchanged

### DISCUSSION THREAD

**Discussion by SimonG85** [PyO3/pyo3#2957] [Questions] [answered]
I have a question.

I'm writing an algorithm that I want to use only in rust but this algo needs some parameters as result of an optimisation.

Since there is a lack of optimisation library in rust (with respect to python) I need to expose some struct in python and some methods with the objective function.

My rust struct
struct Foo <I, S>{
only members that implement IntoPy trait
}

impl Foo {
methods, for example my function to optimise
}

In the python object I need only a concrete implementation of I and S, say i64 and String.

How can I define a struct to wraps Foo into a Python object without changing the rust code (that I need to use after the optimisation)? And I need to export the optimisation function too.

I made a copy of the rust Foo with pyclass and it works but it's a code duplication and I want to avoid if possible.

Thanks.

**Comment by messense** [comment, marked as answer]
Is this what you want?

```rust
use pyo3::prelude::*;

pub struct Foo<I, S> {
    a: I,
    b: S,
}

#[pyclass]
pub struct PyFoo {
    inner: Foo<i64, String>,
}
```

https://www.rustexplorer.com/b/z4p7n1

**Reply by SimonG85** [reply]
OMG, it was easier than expected!
Yes, thanks to your advice I'm able to export python object by using rust struct and function.
In particular I defined the new fuction as follows:

Self {
            inner: Foo::new(...)
        }
where new is a rust function of rust Foo.
Many thanks!

**Reply by SimonG85** [reply]
Another question. In this case I cannot annotate with get and set of pyo3. In this case, do I have to implement getters and setters in rust and expose in python, right?

**Reply by davidhewitt** [reply]
Yes, though I wonder if one day we could have something like

```rust
#[pyclass(getters = (bar = self.inner.bar))]
struct PyFoo {
    inner: Foo
}
```

Might be too powerful 😂

**Reply by SimonG85** [reply]
Yes, indeed. Thanks!
